File: [0126][Easy][Best_Time_to_Buy_and_Sell_Stock_II]/Best_Time_to_Buy_and_Sell_Stock_II_2.java
// solution 2, calculation based on the price peak and valley.
// leetcode time     cost : 1 ms
// leetcode memory   cost : 37.6 MB 
// Time  Complexity: O(N)
// Space Complexity: O(1)
class Solution {
    public int maxProfit(int[] prices) {
        if(prices.length == 0) 
            return 0;
        int i = 0;
        int valley = prices[0];
        int peak = prices[0];
        int maxprofit = 0;
        while (i < prices.length - 1) {
            while (i < prices.length - 1 && prices[i] >= prices[i + 1])
                i++;
            valley = prices[i];
            while (i < prices.length - 1 && prices[i] <= prices[i + 1])
                i++;
            peak = prices[i];
            maxprofit += peak - valley;
        }
        return maxprofit;
    }
}

File: [0122][Easy][Best_Time_to_Buy_and_Sell_Stock_II]/Best_Time_to_Buy_and_Sell_Stock_II_2.py
# solution 3, one time traversal of the price list.
# leetcode time     cost : 72 ms
# leetcode memory   cost : 14.8 MB 
# Time  Complexity: O(N)
# Space Complexity: O(1)
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        return sum([y - x for x, y in zip(prices[:-1], prices[1:]) if x < y])
        #return sum(max(b-a,0)for a,b in zip(prices,prices[1:]))